File: src/libs/m3/com/EPMng.cc
/*
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Init.h>

#include <m3/Syscalls.h>
#include <m3/com/EPMng.h>
#include <m3/tiles/Activity.h>

namespace m3 {

INIT_PRIO_EPMNG EPMng EPMng::_inst;

EP *EPMng::acquire(epid_t ep, uint replies) {
    EP *e = nullptr;
    if(ep == TCU::INVALID_EP && replies == 0)
        e = _eps.remove_first();
    if(!e)
        e = new EP(EP::alloc_for(KIF::SEL_ACT, ep, replies));
    return e;
}

void EPMng::release(EP *ep, bool invalidate) noexcept {
    if(ep->is_standard()) {
        delete ep;
        return;
    }

    if(invalidate) {
        try {
            // invalidate our endpoint to be able to reuse it for something else later
            Syscalls::activate(ep->sel(), ObjCap::INVALID, ObjCap::INVALID, 0);
        }
        catch(...) {
            // ignore errors here
        }
    }

    if(ep->is_cacheable())
        _eps.append(ep);
    else
        delete ep;
}

}

File: src/libs/rust/m3/src/com/epmng.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use col::Vec;
use com::gate::Gate;
use com::{EPArgs, EP};
use errors::Error;
use kif::INVALID_SEL;
use syscalls;

/// The endpoint manager (`EpMng`) multiplexes all non-reserved endpoints among the gates.
pub struct EpMng {
    eps: Vec<EP>,
}

impl EpMng {
    pub fn new() -> Self {
        EpMng { eps: Vec::new() }
    }

    /// Allocates a new endpoint.
    pub fn acquire(&mut self, replies: u32) -> Result<EP, Error> {
        if replies > 0 {
            EP::new_with(EPArgs::new().replies(replies))
        }
        else if let Some(ep) = self.eps.pop() {
            Ok(ep)
        }
        else {
            EP::new()
        }
    }

    /// Frees the given endpoint
    pub fn release(&mut self, ep: EP, invalidate: bool) {
        if invalidate {
            syscalls::activate(ep.sel(), INVALID_SEL, 0).ok();
        }

        if ep.replies() == 0 {
            self.eps.push(ep);
        }
    }

    /// Allocates a new endpoint for the given gate and activates the gate. Returns the endpoint.
    pub(crate) fn activate(&mut self, gate: &Gate) -> Result<EP, Error> {
        let ep = self.acquire(0)?;
        syscalls::activate(ep.sel(), gate.sel(), 0).map(|_| ep)
    }

    pub(crate) fn reset(&mut self) {
        self.eps.clear();
    }
}